src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -269,8 +269,9 @@ private async Task ProcessHeadersFrame(FrameHeader frameHeader)
                 _incomingBuffer.Discard(frameHeader.Length);
 
                 // Don't wait for completion, which could happen asynchronously.
-                ValueTask ignored = SendRstStreamAsync(streamId, Http2ProtocolErrorCode.StreamClosed);
-                return;
+                // TODO: Track the last completed stream ID and include it in the GoAway frame.
+                Task ignored = SendGoAwayAsync(MaxStreamId, Http2ProtocolErrorCode.ProtocolError, new Memory<byte>(Encoding.UTF8.GetBytes("Header frame received on non-open stream.")));
```

Comment:
That's a good point. This pattern makes a lot more sense with SendRstStreamAsync, since it doesn't close the connection.

I can either `await` the task to make sure it sends, or just strip out the GoAway code since that isn't a strict requirement in the spec and is mostly just a debugging thing. I don't really ahve a preference between the two.

Reply:
If we do decide we want to try to send a GOAWAY in these cases (see my comments below), then I think we should change where the GOAWAY is sent. It should happen as part of the abort logic that's initiated in ProcessIncomingFramesAsync when an exception (specifically, a, Http2ProtocolException) is caught. This avoids the race here between SendGoAwayAsync and the connection close in the abort logic; we would await the former before doing the latter. 

It won't avoid a potential connection reset due to other issues (see my comments below).

Reply:
Doing this would also guarantee we send GOAWAY in other protocol error cases, not just these specific cases.